<a href="https://colab.research.google.com/github/AI-KPI/ml_intro/blob/main/linear_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Plot style
sns.set()
%pylab inline
pylab.rcParams['figure.figsize'] = (4, 4)

Populating the interactive namespace from numpy and matplotlib


In [3]:
%%html
<style>
.pquote {
  text-align: left;
  margin: 40px 0 40px auto;
  width: 70%;
  font-size: 1.5em;
  font-style: italic;
  display: block;
  line-height: 1.3em;
  color: #5a75a7;
  font-weight: 600;
  border-left: 5px solid rgba(90, 117, 167, .1);
  padding-left: 6px;
}
.notes {
  font-style: italic;
  display: block;
  margin: 40px 10%;
}
img + em {
  text-align: center;
  display: block;
  color: gray;
  font-size: 0.9em;
  font-weight: 600;
}
</style>

$$
\newcommand\norm[1]{\left\lVert#1\right\rVert} 
\DeclareMathOperator{\Tr}{Tr}
\newcommand\bs[1]{\boldsymbol{#1}}
$$

<span class='notes'>
    Цей матеріал є частиною серії 2 глави про лінійну алгебру з [Deep Learning Book](http://www.deeplearningbook.org/) by Goodfellow, I., Bengio, Y., and Courville, A. (2016). Він спрямований на надання інформації, ілюстрацій та коду на мові python, про математичні теорії і побудований на моєму розуміння цих понять. Ви можете перевірити навчальний план у [introduction post(вступній публікації)] (https://hadrienj.github.io/posts/Deep-Learning-Book-Series-Introduction/).
</span>

# Вступ

Цей розділ дуже легкий! Я можу запевнити вас, що ви прочитаєте це за 1 хвилину! Приємно після останніх двох розділів, які були досить великими! Ми побачимо, що таке слід матриці. Це знадобиться для останнього розділу аналізу основних компонентів (PCA).

# 2.10 Оператор Слід 

<img src="images/trace-matrix.png" width="200" alt="Обчислення сліду матриці" title="Calculating the trace of a matrix">
<em>Слід матриці</em>


Слід - це сума всіх значень по діагоналі квадратної матриці.

$$
\bs{A}=
\begin{bmatrix}
    2 & 9 & 8 \\\\
    4 & 7 & 1 \\\\
    8 & 2 & 5
\end{bmatrix}
$$

$$
\mathrm{Tr}(\bs{A}) = 2 + 7 + 5 = 14
$$

Numpy надає функцію `trace ()` для її обчислення:

In [4]:
A = np.array([[2, 9, 8], [4, 7, 1], [8, 2, 5]])
A

array([[2, 9, 8],
       [4, 7, 1],
       [8, 2, 5]])

In [5]:
A_tr = np.trace(A)
A_tr

14

GoodFellow та ін. пояснили, що слід можна використовувати для вказівки норми Фробеніуса для матриці (переглянте [2.5](https://hadrienj.github.io/posts/Deep-Learning-Book-Series-2.5-Norms/)). Норма Фробениуса є еквівалентом норми $L^2$ для матриць. Він визначається:

$$
\norm{\bs{A}}_F=\sqrt{\sum_{i,j}A^2_{i,j}}
$$

Візьміть квадрат усіх елементів і підсумуйте їх. Візьміть квадратний корінь результату. Цю норму також можна обчислити за:

$$
\norm{\bs{A}}_F=\sqrt{\Tr({\bs{AA}^T})}
$$

Ми можемо це перевірити. Перший спосіб обчислення норми можна зробити за допомогою простої команди `np.linalg.norm()`:

In [6]:
np.linalg.norm(A)

17.549928774784245

Норма Фробениуса $\bs{A}$ is 17.549928774784245.

Зі слідом результат ідентичний:

In [7]:
np.sqrt(np.trace(A.dot(A.T)))

17.549928774784245

Оскільки транспозиція матриці не змінює діагоналі, слід матриці дорівнює сліду її транспонування:

$$
\Tr(\bs{A})=\Tr(\bs{A}^T)
$$

## Слід добутку

$$
\Tr(\bs{ABC}) = \Tr(\bs{CAB}) = \Tr(\bs{BCA})
$$


### Приклад 1.

Подивимось приклад цієї властивості.

$$
\bs{A}=
\begin{bmatrix}
    4 & 12 \\\\
    7 & 6
\end{bmatrix}
$$

$$
\bs{B}=
\begin{bmatrix}
    1 & -3 \\\\
    4 & 3
\end{bmatrix}
$$

$$
\bs{C}=
\begin{bmatrix}
    6 & 6 \\\\
    2 & 5
\end{bmatrix}
$$

In [8]:
A = np.array([[4, 12], [7, 6]])
B = np.array([[1, -3], [4, 3]])
C = np.array([[6, 6], [2, 5]])

np.trace(A.dot(B).dot(C))

531

In [9]:
np.trace(C.dot(A).dot(B))

531

In [10]:
np.trace(B.dot(C).dot(A))

531

$$
\bs{ABC}=
\begin{bmatrix}
    360 & 432 \\\\
    180 & 171
\end{bmatrix}
$$

$$
\bs{CAB}=
\begin{bmatrix}
    498 & 126 \\\\
    259 & 33
\end{bmatrix}
$$

$$
\bs{BCA}=
\begin{bmatrix}
    -63 & -54 \\\\
    393 & 594
\end{bmatrix}
$$

$$
\Tr(\bs{ABC}) = \Tr(\bs{CAB}) = \Tr(\bs{BCA}) =  531
$$

<span class='notes'>
    Не соромтеся писати мені електронний лист чи коментар. Програму цієї серії можна знайти [in the introduction post(у вступній статті)](https://hadrienj.github.io/posts/Deep-Learning-Book-Series-Introduction/). Всі зошити можна знайти на [Github](https://github.com/hadrienj/deepLearningBook-Notes).
</span>

# Список літератури

[Trace (linear algebra) - Wikipedia (Слід (лінійна алгебра) - Вікіпедія)](https://en.wikipedia.org/wiki/Trace_(linear_algebra))

[Numpy Trace operator(Оператор Numpy Trace)](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trace.html)

Переклад: Якимів Іван, студент ІО-02мп, 2020 рік